ethereum/ratelimit/rate_limiter_test.go

```diff
@@ -0,0 +1,234 @@
+package ratelimit
+
+import (
+	"context"
+	"math"
+	"testing"
+	"time"
+
+	"github.com/0xProject/0x-mesh/constants"
+	"github.com/0xProject/0x-mesh/meshdb"
+	"github.com/benbjohnson/clock"
+	"github.com/google/uuid"
+	"github.com/stretchr/testify/assert"
+	"github.com/stretchr/testify/require"
+)
+
+var (
+	maxRequestsPer24HrsWithoutBuffer = 301000
+	maxRequestsPer24hrs              = maxRequestsPer24HrsWithoutBuffer - maxRequestsPer24HrsBuffer
+	maxRequestsPerSecond             = 10.0
+	twentyFourHrs                    = (24 * 60 * 60 * 1000 * time.Millisecond)
+	maxExpectedDelay                 = twentyFourHrs / time.Duration(maxRequestsPer24hrs)
+	minExpectedDelay                 = time.Duration(1000) / time.Duration(maxRequestsPerSecond) * time.Millisecond
+	defaultCheckpointInterval        = 1 * time.Minute
+)
+
+// Scenario1: Mesh starts X seconds after UTC midnight (start of next UTC day) and
+// therefore there are Y request grants that have accrued. This test verifies that
+// the first request is granted immediately, the next Y - 1 grants are issued at the
+// expected minimum rate imposed by the per second rate, and all subsequent requests
+// are issued at the max rate imposed by the per 24hr rate limit.
+func TestScenario1(t *testing.T) {
+	meshDB, err := meshdb.New("/tmp/meshdb_testing/" + uuid.New().String())
+	require.NoError(t, err)
+	defer meshDB.Close()
+
+	initMetadata(t, meshDB)
+
+	// Set mock clock to three grants past UTC midnight
+	aClock := clock.NewMock()
+	now := time.Now()
+	midnightUTC := getUTCMidnightOfDate(now)
+	threeGrantsPastUTCMidnight := midnightUTC.Add(maxExpectedDelay * 3)
+	aClock.Set(threeGrantsPastUTCMidnight)
+
+	rateLimiter, err := New(maxRequestsPer24HrsWithoutBuffer, maxRequestsPerSecond, meshDB, aClock)
+	require.NoError(t, err)
+	ctx, cancel := context.WithCancel(context.Background())
+	defer cancel()
+	go func() {
+		err := rateLimiter.Start(ctx, defaultCheckpointInterval)
+		require.NoError(t, err)
+	}()
+
+	for i := 0; i < 5; i++ {
+		now := time.Now()
+		ctx := context.Background()
+		err = rateLimiter.Wait(ctx)
+		require.NoError(t, err)
+		elapsed := time.Since(now)
+
+		// First request goes through immediately
+		if i == 0 {
+			assert.Condition(t, func() bool {
+				return elapsed < 1*time.Millisecond
+			})
+		} else if i > 0 && i <= 3 {
+			// Subsequent requests take 1sec / maxRequestsPerSecond
+			// Note: Despite initially waiting for 3 grants to accrue, by
+			// the time we request the 4th, another has accrued.
+			delta := math.Abs(float64(minExpectedDelay - elapsed))
+			assert.Condition(t, func() bool {
+				return time.Duration(delta) < 5*time.Millisecond
```

Comment:
Can we use mathematical expression based on the constants above instead of hard-coding `5 * time.Millisecond`?

Reply:
There is naturally some variation in the delta between the `minExpectedDelay` and the `elapsed` time. I think adding a constant for this delta is fine.

Reply:
What I'm worried about is that this will unexpectedly break if we ever change the constants defined above.

Reply:
I would also be okay with extracting these values out and declaring them as constants right next to the other ones we already have. That way if we change one of the constants we have a better idea of where to look if something breaks.